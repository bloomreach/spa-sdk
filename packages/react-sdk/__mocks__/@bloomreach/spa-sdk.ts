/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const spaSdk = jest.requireActual('@bloomreach/spa-sdk');

export const META_POSITION_BEGIN = spaSdk.META_POSITION_BEGIN;
export const META_POSITION_END = spaSdk.META_POSITION_END;

export function mockMeta(data: string, position: string) {
  return {
    getData: jest.fn().mockReturnValue(data),
    getPosition: jest.fn().mockReturnValue(position),
  };
}

export function mockNoCommentMeta() {
  return mockMeta('not-a-comment', META_POSITION_BEGIN);
}

const componentMock = {
  getMeta: jest.fn().mockReturnValue([
    mockMeta('meta-begin', META_POSITION_BEGIN),
    mockMeta('meta-end', META_POSITION_END),
  ]),
};

export const pageMock = {
  getComponent: jest.fn().mockReturnValue(componentMock),
  getContent: jest.fn(),
  getTitle: jest.fn(),
  sync: jest.fn(),
};

export const initialize = jest.fn();
export const destroy = jest.fn();

export function isMetaComment(value: any) {
  return value.getData() !== 'not-a-comment';
}
