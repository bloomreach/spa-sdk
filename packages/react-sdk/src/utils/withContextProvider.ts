// eslint-disable-next-line import/no-extraneous-dependencies
import PropTypes from 'prop-types';
import React, { ReactElement } from 'react';

export const withContextProvider = (context?: Record<string, any>, children?: ReactElement): React.ReactElement => {
  const childContextTypes = (): Record<string, string> | undefined => context
    && Object.keys(context).reduce((obj: Record<string, any>, key: string) => {
      obj[key] = PropTypes.any;
      return obj;
    }, {});

  class ContextProvider extends React.Component {
    getChildContext(): Record<string, string> | undefined {
      return context;
    }

    render(): React.ReactNode {
      return children;
    }
  }

  (ContextProvider as any).displayName = 'ContextProvider';
  (ContextProvider as any).childContextTypes = childContextTypes();

  return React.createElement(ContextProvider, null);
};
