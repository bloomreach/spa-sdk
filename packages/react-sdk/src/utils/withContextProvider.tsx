/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement } from 'react';
import { BrPageContext } from '../page';
import { BrComponentContext, BrMappingContext } from '../component';

export const withPageContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return <BrPageContext.Provider value={context}>{children}</BrPageContext.Provider>;
};

export const withMappingContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return (
    <BrMappingContext.Provider value={context}>
      {children}
    </BrMappingContext.Provider>
  );
};

export const withComponentContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return <BrComponentContext.Provider value={context}>{children}</BrComponentContext.Provider>;
};
