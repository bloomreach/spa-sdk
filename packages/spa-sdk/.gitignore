/coverage
/docs
/lib
/node_modules
.DS_store