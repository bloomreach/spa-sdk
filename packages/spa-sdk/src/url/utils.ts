/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

interface Url {
  hash: string;
  origin: string;
  pathname: string;
  search: string;
  searchParams: URLSearchParams;
  path: string;
}

export function buildUrl(url: Partial<Url>) {
  const searchParams = url.searchParams?.toString() ?? '';
  const search = url.search ?? `${searchParams && `?${searchParams}`}`;
  const path = url.path ?? `${url.pathname ?? ''}${search}${url.hash ?? ''}`;

  return `${url.origin ?? ''}${path}`;
}

function isMatchedOrigin(origin: string, baseOrigin: string) {
  return !baseOrigin || !origin || baseOrigin === origin;
}

function isMatchedPathname(pathname: string, basePathname: string) {
  return !basePathname || pathname.startsWith(basePathname);
}

function isMatchedQuery(search: URLSearchParams, baseSearch: URLSearchParams) {
  let match = true;
  baseSearch.forEach((value, key) => {
    match = match && (!value && search.has(key) || search.getAll(key).includes(value));
  });

  return match;
}

export function isMatched(link: string, base = '') {
  const linkUrl = parseUrl(link);
  const baseUrl = parseUrl(base);

  return isMatchedOrigin(linkUrl.origin, baseUrl.origin)
    && isMatchedPathname(linkUrl.pathname, baseUrl.pathname)
    && isMatchedQuery(linkUrl.searchParams, baseUrl.searchParams);
}

export function mergeSearchParams(params: URLSearchParams, ...rest: URLSearchParams[]) {
  const result = new URLSearchParams(params);
  rest.forEach(params => params.forEach((value, key) => result.set(key, value)));

  return result;
}

export function parseUrl(url: string): Url {
  // URL constructor requires either a valid URL or a base URL.
  // Since this function returns a pathname, we can safely pass a fake host to be able to resolve relative URLs.
  const parsedUrl = url ? new URL(url, 'http://example.com') : {} as URL;
  const { hash = '', search = '', searchParams = new URLSearchParams() } = parsedUrl;

  // For links like `//example.com?query#hash` pathname will be `/` so we need to strip query and hash parameters first.
  let origin = url.substring(0, url.length - search.length - hash.length);
  let { pathname = '' } = parsedUrl;
  if (!origin.endsWith(pathname)) {
    pathname = pathname.substring(1);
  }

  origin = origin.substring(0, origin.length - pathname.length);

  return { hash, origin, pathname, search, searchParams, path: `${pathname}${search}${hash}` };
}
