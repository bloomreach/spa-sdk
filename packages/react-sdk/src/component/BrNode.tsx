/*
 * Copyright 2019-2020 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { isContainer, isContainerItem, Component } from '@bloomreach/spa-sdk';
import { BrMeta } from '../meta';
import { BrPageContext } from '../page/BrPageContext';
import { BrComponentContext } from './BrComponentContext';
import { BrNodeContainer } from './BrNodeContainer';
import { BrNodeContainerItem } from './BrNodeContainerItem';
import { BrNodeComponent } from './BrNodeComponent';

interface BrNodeProps {
  component?: Component;
}

export class BrNode extends React.Component<BrNodeProps> {
  static contextType = BrPageContext;

  context: React.ContextType<typeof BrPageContext>;

  private renderNode(): JSX.Element | React.ReactNode {
    const { children, component } = this.props;

    if (React.Children.count(children)) {
      return <BrMeta meta={component?.getMeta()}>{children}</BrMeta>;
    }

    // eslint-disable-next-line react/no-array-index-key
    const childrenList = component?.getChildren().map((child, index) => <BrNode key={index} component={child} />);

    if (isContainer(component)) {
      return (
        <BrNodeContainer component={component} page={this.context!}>
          {childrenList}
        </BrNodeContainer>
      );
    }

    if (isContainerItem(component)) {
      return (
        <BrNodeContainerItem component={component} page={this.context!}>
          {childrenList}
        </BrNodeContainerItem>
      );
    }

    return (
      <BrNodeComponent component={component} page={this.context!}>
        {childrenList}
      </BrNodeComponent>
    );
  }

  render(): JSX.Element {
    const { component } = this.props;

    return <BrComponentContext.Provider value={component}>{this.renderNode()}</BrComponentContext.Provider>;
  }
}
