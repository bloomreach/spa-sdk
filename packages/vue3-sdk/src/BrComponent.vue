<!--
  Copyright 2023 Bloomreach

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

    https://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<template>
  <br-node-component v-for="component in components" :component="component">
    <slot :component="component" :page="page"></slot>
  </br-node-component>
</template>

<script setup lang="ts">
import BrNodeComponent from '@/BrNodeComponent.vue';
import { component$, page$ } from '@/providerKeys';
import type { Component } from '@bloomreach/spa-sdk';
import { isComponent } from '@bloomreach/spa-sdk';
import { computed, inject } from 'vue';

function getComponents(
  parentComponent: Component | undefined,
  propsComponent: Component | string | undefined,
): Component[] {
  if (isComponent(propsComponent)) {
    return [propsComponent];
  }

  if (!parentComponent) {
    return [];
  }

  if (!propsComponent) {
    return parentComponent.getChildren();
  }

  const parentComponents = parentComponent.getComponent(...propsComponent.split('/'));
  return parentComponents ? [parentComponents] : [];
}

const props = defineProps<{ component?: Component | string }>();
const page = inject(page$);
const parent = inject(component$);
const components = computed(() => getComponents(parent?.value, props.component));
</script>
