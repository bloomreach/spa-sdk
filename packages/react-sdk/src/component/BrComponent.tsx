/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { isComponent, Component } from '@bloomreach/spa-sdk';
import { BrComponentContext } from './BrComponentContext';
import { BrNode } from './BrNode';

interface BrComponentProps {
  path?: string;
}

interface BrComponentState {
  components: Component[];
}

export class BrComponent extends React.Component<BrComponentProps, BrComponentState> {
  static contextType = BrComponentContext;

  constructor(props: BrComponentProps, public context: React.ContextType<typeof BrComponentContext>) {
    super(props, context);

    this.state = {
      components: this.getComponents(),
    };
  }

  componentDidUpdate(prevProps: BrComponentProps) {
    if (this.props.path !== prevProps.path) {
      this.setState({ components: this.getComponents() });
    }
  }

  private getComponents() {
    if (!this.context) {
      return [];
    }
    if (!this.props.path) {
      return this.context.getChildren();
    }

    const component = this.context.getComponent(...this.props.path.split('/'));

    return component ? [component] : [];
  }

  private renderComponents() {
    return this.state.components.map((component, index) => (
      <BrNode key={index} component={component}>{this.props.children}</BrNode>
    ));
  }

  render() {
    return <>{this.renderComponents()}</>;
  }
}
