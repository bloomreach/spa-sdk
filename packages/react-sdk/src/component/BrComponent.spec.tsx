/*
 * Copyright 2019-2022 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Component } from '@bloomreach/spa-sdk';
import { render } from '@testing-library/react';
import { BrComponent } from './BrComponent';
import { BrNode } from './BrNode';
import { withContextProvider } from '../utils/withContextProvider';

jest.mock('@bloomreach/spa-sdk');

describe('BrComponent', () => {
  const context = {
    getChildren: jest.fn(() => []),
    getMeta: jest.fn(),
    getComponent: jest.fn(),
  } as unknown as jest.Mocked<Component>;
  const component = {
    getChildren: jest.fn(() => []),
    getMeta: jest.fn(),
    getName: jest.fn(),
  } as unknown as Component;

  beforeEach(() => {
    jest.restoreAllMocks();

    (BrComponent as any).contextTypes = {
      getChildren: () => null,
      getMeta: () => null,
      getComponent: () => null,
    };
    delete (BrComponent as Partial<typeof BrComponent>).contextType;
  });

  it('should render nothing if there is no context', () => {
    delete (BrComponent as any).contextTypes;
    const element = render(<BrComponent />);

    expect(element.container.firstChild).toBe(null);
  });

  it('should render children if path is not set', () => {
    const component1 = { ...component } as Component;
    const component2 = { ...component } as Component;
    context.getChildren.mockReturnValueOnce([component1, component2]);
    const element = render(withContextProvider(context, <BrComponent />));

    expect(context.getChildren).toBeCalled();

    const node1 = render(withContextProvider(context, <BrNode component={component1} />));
    const node2 = render(withContextProvider(context, <BrNode component={component2} />));

    expect(element.container).toContainHTML(node1.container.innerHTML);
    expect(element.container).toContainHTML(node2.container.innerHTML);
  });

  it('should split path by slashes', () => {
    render(withContextProvider(context, <BrComponent path="a/b" />));

    expect(context.getComponent).toBeCalledWith('a', 'b');
  });

  it('should render nothing if no component found', () => {
    const element = render(withContextProvider(context, <BrComponent path="a/b" />));

    expect(element.container.firstChild).toBe(null);
    expect(element.asFragment()).toMatchSnapshot();
  });

  it('should render found component', () => {
    context.getComponent.mockReturnValueOnce(component);
    const element = render(withContextProvider(context, <BrComponent path="a/b" />));

    const node = render(withContextProvider(context, <BrNode component={component} />));

    expect(element.container).toContainHTML(node.container.innerHTML);
  });

  it('should pass children down', () => {
    context.getComponent.mockReturnValueOnce(component);
    const element = render(
      withContextProvider(
        context,
        <BrComponent path="a/b">
          <a />
        </BrComponent>,
      ),
    );
    const node = render(
      withContextProvider(
        context,
        <BrNode component={component}>
          <a />
        </BrNode>,
      ),
    );

    expect(element.container).toContainHTML(node.container.innerHTML);
  });
});
