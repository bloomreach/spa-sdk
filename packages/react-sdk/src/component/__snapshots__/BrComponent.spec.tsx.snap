// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrComponent should render nothing if no component found 1`] = `<DocumentFragment />`;
